e2e/tests/wasm/grandpa_test.go

```diff
@@ -255,6 +257,160 @@ func (s *GrandpaTestSuite) TestMsgTransfer_Succeeds_GrandpaContract() {
+
+	s.Require().NotEmpty(codeHash, "codehash was empty but should not have been")
+
+	eRep := s.GetRelayerExecReporter()
+
+	// Set client contract hash in cosmos chain config
+	err = r.SetClientContractHash(ctx, eRep, cosmosChain.Config(), codeHash)
+	s.Require().NoError(err)
+
+	// Ensure parachain has started (starts 1 session/epoch after relay chain)
+	err = testutil.WaitForBlocks(ctx, 1, polkadotChain)
+	s.Require().NoError(err, "polkadot chain failed to make blocks")
+
+	// Fund users on both cosmos and parachain, mints Asset 1 for Alice
+	fundAmount := int64(12_333_000_000_000)
+	polkadotUser, cosmosUser := s.fundUsers(ctx, fundAmount, polkadotChain, cosmosChain)
+
+	_ = polkadotUser
+	_ = cosmosUser
+
+	pathName := s.GetPathName(0)
+
+	err = r.GeneratePath(ctx, eRep, cosmosChain.Config().ChainID, polkadotChain.Config().ChainID, pathName)
+	s.Require().NoError(err)
+
+	// create client pair with subject (bad trusting period)
+	subjectClientID := clienttypes.FormatClientIdentifier(ibcexported.Wasm, 0)
+	// TODO(damian): The hyperspace relayer makes no use of create client opts
+	// https://github.com/strangelove-ventures/interchaintest/blob/main/relayer/hyperspace/hyperspace_commander.go#L83
+	s.SetupClients(ctx, r, ibc.CreateClientOptions{
+		TrustingPeriod: badTrustingPeriod.String(), // NOTE: this is hardcoded within the cw contract: ics10_grapnda_cw_expiry.wasm
+	})
+
+	// wait for block
+	err = testutil.WaitForBlocks(ctx, 1, cosmosChain, polkadotChain)
+	s.Require().NoError(err)
+
+	// wait the bad trusting period
+	time.Sleep(badTrustingPeriod)
+
+	// create client pair with substitue
+	substituteClientID := clienttypes.FormatClientIdentifier(ibcexported.Wasm, 1)
+	s.SetupClients(ctx, r, ibc.DefaultClientOpts())
+
+	// wait for block
+	err = testutil.WaitForBlocks(ctx, 1, cosmosChain, polkadotChain)
+	s.Require().NoError(err)
+
+	// ensure subject client is expired
+	// TODO(damian): The grandpa-cw contract does not seem to have a concept of trusting period.
+	// https://github.com/ComposableFi/composable-ibc/blob/master/light-clients/ics10-grandpa-cw/src/contract.rs#L343-L357
+	status, err := s.clientStatus(ctx, cosmosChain, subjectClientID)
+	s.Require().NoError(err)
+	s.Require().Equal(ibcexported.Expired.String(), status, "unexpected subject client status")
+
+	// ensure substitute client is active
+	status, err = s.clientStatus(ctx, cosmosChain, substituteClientID)
+	s.Require().NoError(err)
+	s.Require().Equal(ibcexported.Active.String(), status, "unexpected substitute client status")
+
+	// create and execute a client recovery proposal
+	authority, err := s.QueryModuleAccountAddress(ctx, govtypes.ModuleName, cosmosChain)
+	s.Require().NoError(err)
+	msgRecoverClient := clienttypes.NewMsgRecoverClient(authority.String(), subjectClientID, substituteClientID)
+	s.Require().NotNil(msgRecoverClient)
+	s.ExecuteAndPassGovV1Proposal(ctx, msgRecoverClient, cosmosChain, cosmosUser)
```

Comment:
This test running with the contract mentioned in https://github.com/cosmos/ibc-go/pull/5027/files#r1386287352 is failing here, when the `MsgRecoverClient` is delivered by gov. It calls through ibc core and into the wasmvm returning the following:

```
client recovery failed: failed to validate substitute client: Error calling the VM: Error executing Wasm: Wasmer runtime error: RuntimeError: Error calling into the VM's backend: Panic in FFI call: wasm contract call failed: failed waiting for condition after 300.000000 seconds
```

Prior to this error being returned (before @misko9's branch was rebased with main of the upstream) I could get as far as this point: https://github.com/misko9/centauri/blob/steve/contracts/light-clients/ics10-grandpa-cw/src/contract.rs#L259 but also yielded a failure for the equality check:

```
client recovery failed: failed to validate substitute client: Grandpa error: subject client state does not match substitute client state: wasm contract call failed: failed waiting for condition after 300.000000 seconds
```